cmd/relay/forward.go

```diff
@@ -0,0 +1,91 @@
+package main
+
+import (
+	"bytes"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/bluesky-social/indigo/cmd/relay/relay"
+
+	"github.com/labstack/echo/v4"
+)
+
+// Forwards HTTP request on to sibling relay instances, if they are configured.
+//
+// This method expects to be run "in the background" as a goroutine, so it doesn't take a `context.Context`, and does not return an `error`. It logs both success and failure. The `echo.Context` request has presumably been read, so any body is passed separately. The `echo.Context` response is presumably being returned or already finalized concurrently.
+func (s *Service) ForwardSiblingRequest(c echo.Context, body []byte) {
+
+	if len(s.config.SiblingRelayHosts) == 0 {
+		return
+	}
+
+	// if the request itself was already forwarded, or user-agent seems to be a relay, then don't forward on further (to prevent loops)
+	req := c.Request()
+	for _, via := range req.Header.Values("Via") {
+		if strings.Contains(via, "atproto-relay") {
+			s.logger.Info("not re-forwarding request to sibling relay", "header", "Via", "value", via)
+			return
+		}
+	}
+	for _, ua := range req.Header.Values("User-Agent") {
+		if strings.Contains(ua, "atproto-relay") {
+			s.logger.Info("not re-forwarding request to sibling relay", "header", "User-Agent", "value", ua)
+			return
+		}
+	}
+
+	// TODO: could turn this in to a shared/persistent client
+	client := http.Client{
+		Timeout: 10 * time.Second,
+	}
+
+	for _, rawHost := range s.config.SiblingRelayHosts {
+		hostname, noSSL, err := relay.ParseHostname(rawHost)
+		if err != nil {
+			s.logger.Error("invalid sibling hostname configured", "host", rawHost, "err", err)
+			return
+		}
+		u := req.URL
+		u.Host = hostname
+		if noSSL {
+			u.Scheme = "http"
+		} else {
+			u.Scheme = "https"
+		}
+		var b io.Reader
+		if body != nil {
+			b = bytes.NewBuffer(body)
+		}
+		upstreamReq, err := http.NewRequest(req.Method, u.String(), b)
+		if err != nil {
+			s.logger.Error("creating admin forward request failed", "method", req.Method, "url", u.String(), "err", err)
+			continue
+		}
+
+		// copy some headers from inbound request
+		for _, hdr := range []string{"Accept", "User-Agent", "Authorization", "Content-Type"} {
+			val := req.Header.Get(hdr)
+			if val != "" {
+				upstreamReq.Header.Set(hdr, val)
+			}
+		}
+
+		// add Via header (critical to prevent forwarding loops)
+		upstreamReq.Header.Add("Via", s.relay.Config.UserAgent)
```

Comment:
Two little things, not critical in any way:
 - we might append to the `Via` from the incoming request.
 - each item in the `Via` header has two parts: first the http version, e.g. `1.1`, and then second a name not containing whitespace.  An example might be `1.1 bsky-relay1-us-east, 1.1 bsky-relay2-us-west`.

Reply:
Good catch on the `Via` header, I had looked it up but forgot: https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Via

I went with the full proto, followed by the pseudoname "atproto-relay", for example "HTTP/1.1 atproto-relay". To prevent loops it is critical for this to reliably have a detectable sub-string (like "atproto-relay"), so I went with that over config or trying to auto-detect.

The stdlib `Header.Add()` method does append to any existing values (unlike `Header.Set()`); and the code above which checks headers uses `Header.Values()` to iterate over all values.

Reply:
Nice!  

Re: adding vs setting the header, my understanding was that it was adding it but that there were no values set from the original incoming `req`.  It is probably alright either way for our purposes 👍